File: gd_acc/gd_accomodation/report/bed_status_history_report/bed_status_history_report.py
# Copyright (c) 2026, Rahmed-dev and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import getdate

HISTORY_FIELDS = (
	"change_datetime",
	"bed",
	"room",
	"floor",
	"site",
	"previous_status",
	"new_status",
	"reason",
	"employee",
	"employee_name",
	"allocation",
	"transfer",
	"changed_by",
)


def execute(filters=None):
	filters = frappe._dict(filters or {})
	return get_columns(), get_data(filters)


def get_columns():
	return [
		{
			"label": _("Changed On"),
			"fieldname": "change_datetime",
			"fieldtype": "Datetime",
			"width": 170,
		},
		{
			"label": _("Bed"),
			"fieldname": "bed",
			"fieldtype": "Link",
			"options": "Accommodation Bed",
			"width": 180,
		},
		{
			"label": _("Room"),
			"fieldname": "room",
			"fieldtype": "Link",
			"options": "Accommodation Room",
			"width": 160,
		},
		{
			"label": _("Floor"),
			"fieldname": "floor",
			"fieldtype": "Link",
			"options": "Accommodation Floor",
			"width": 160,
		},
		{
			"label": _("Accommodation Site"),
			"fieldname": "site",
			"fieldtype": "Link",
			"options": "Accommodation Site",
			"width": 180,
		},
		{
			"label": _("Previous Status"),
			"fieldname": "previous_status",
			"fieldtype": "Data",
			"width": 130,
		},
		{
			"label": _("New Status"),
			"fieldname": "new_status",
			"fieldtype": "Data",
			"width": 130,
		},
		{
			"label": _("Reason"),
			"fieldname": "reason",
			"fieldtype": "Data",
			"width": 130,
		},
		{
			"label": _("Employee"),
			"fieldname": "employee",
			"fieldtype": "Link",
			"options": "Employee",
			"width": 130,
		},
		{
			"label": _("Employee Name"),
			"fieldname": "employee_name",
			"fieldtype": "Data",
			"width": 180,
		},
		{
			"label": _("Allocation"),
			"fieldname": "allocation",
			"fieldtype": "Link",
			"options": "Accommodation Allocation",
			"width": 170,
		},
		{
			"label": _("Transfer"),
			"fieldname": "transfer",
			"fieldtype": "Link",
			"options": "Accommodation Transfer",
			"width": 160,
		},
		{
			"label": _("Changed By"),
			"fieldname": "changed_by",
			"fieldtype": "Link",
			"options": "User",
			"width": 160,
		},
	]


def get_data(filters):
	return frappe.get_all(
		"Bed Status History",
		filters=get_history_filters(filters),
		fields=list(HISTORY_FIELDS),
		order_by="change_datetime desc, creation desc",
	)


def get_history_filters(filters):
	history_filters = []

	for fieldname in ("bed", "location", "site", "reason"):
		if filters.get(fieldname):
			history_filters.append([fieldname, "=", filters.get(fieldname)])

	if filters.get("from_date"):
		history_filters.append(["change_datetime", ">=", f"{getdate(filters.from_date)} 00:00:00"])
	if filters.get("to_date"):
		history_filters.append(["change_datetime", "<=", f"{getdate(filters.to_date)} 23:59:59"])

	return history_filters

File: gd_acc/gd_accomodation/report/bed_status_history_report/bed_status_history_report.js
// Copyright (c) 2026, Rahmed-dev and contributors
// For license information, please see license.txt

frappe.query_reports["Bed Status History Report"] = {
	filters: [
		{
			fieldname: "location",
			label: __("Location"),
			fieldtype: "Link",
			options: "Accommodation Location",
		},
		{
			fieldname: "site",
			label: __("Accommodation Site"),
			fieldtype: "Link",
			options: "Accommodation Site",
			get_query: function () {
				const filters = {};
				const location = frappe.query_report.get_filter_value("location");
				if (location) {
					filters.location = location;
				}
				return { filters: filters };
			},
		},
		{
			fieldname: "bed",
			label: __("Bed"),
			fieldtype: "Link",
			options: "Accommodation Bed",
			get_query: function () {
				const filters = {};
				const location = frappe.query_report.get_filter_value("location");
				const site = frappe.query_report.get_filter_value("site");
				if (location) {
					filters.location = location;
				}
				if (site) {
					filters.site = site;
				}
				return { filters: filters };
			},
		},
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
		},
		{
			fieldname: "reason",
			label: __("Reason"),
			fieldtype: "Select",
			options: "\nAllocation\nRelease\nTransfer In\nTransfer Out\nManual Update\nMaintenance\nBulk Setup\nHold\nHold Released",
		},
	],

	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		const status_colors = {
			Available: "green",
			Occupied: "blue",
			Reserved: "orange",
			Maintenance: "red",
			Blocked: "red",
			Inactive: "gray",
		};

		if (column.fieldname === "new_status" && data && data.new_status) {
			value = `<span class="indicator-pill ${status_colors[data.new_status] || "gray"}">${__(
				data.new_status
			)}</span>`;
		}

		return value;
	},
};
